### PR TITLE
enh(platform topology): Registration script rework

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

This PR replace the PHP registration script by a Bash Version.

Need to be done:
 - Adapt documentation to use the .sh script

**Fixes** # MON-6432

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.10.x
- [ ] 20.04.x
- [x] 20.10.x
- [x] 21.04.x (master)

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: bin/registerServerTopology.sh [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+###########################################################
+
+#========= begin of function parse_command_options()
+# This function will parse the flag passed to the command and assign them to variables
+# If mandatories options are missing an error is returned
+function parse_command_options() {
+  while (($# > 0)); do
+    case $1 in
+        -t|--type)
+          set_variable "CURRENT_NODE_TYPE" "$2"
```
Comment (sc979) [inline bin/registerServerTopology.sh:32]:
```suggestion
          CURRENT_NODE_TYPE="$2"
```

File: bin/registerServerTopology.sh [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+          # TODO: If Remote call endpoint to register remote
+          shift 2
```
Comment (sc979) [inline bin/registerServerTopology.sh:34]: won't shift 2 close the while loop ?

File: bin/registerServerTopology.sh [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+}
+#========= end of function set_variable()
+
+
+#========= begin of function get_api_token()
+# Get the X-AUTH-TOKEN used in register request
+function get_api_token() {
+  API_RESPONSE=$(curl -s -X POST ${INSECURE:+--insecure} -H "Content-Type: application/json" \
+    -d '{"security":{"credentials":{"login":"'"${API_USERNAME}"'", "password":"'"$1"'"}}}' \
+    "${TARGET_NODE_ADDRESS}/${ROOT_CENTREON_FOLDER}/api/latest/login")
```
Comment (sc979) [inline bin/registerServerTopology.sh:111]:
```suggestion
  API_RESPONSE=`$(curl -s -X POST ${INSECURE:+--insecure} -H "Content-Type: application/json" \
    -d '{"security":{"credentials":{"login":"'"${API_USERNAME}"'", "password":"'"$1"'"}}}' \
    "${TARGET_NODE_ADDRESS}/${ROOT_CENTREON_FOLDER}/api/latest/login")`
```

File: bin/registerServerTopology.sh [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+
+  API_TOKEN=$( echo "${API_RESPONSE}" | grep -o '"token":"[^"]*' | cut -d'"' -f4)
```
Comment (sc979) [inline bin/registerServerTopology.sh:113]:
```suggestion
  API_TOKEN=`$( echo "${API_RESPONSE}" | grep -o '"token":"[^"]*' | cut -d'"' -f4)`
```

File: bin/registerServerTopology.sh [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+}
+#========= end of function get_api_token()
+
+
+#========= begin of function log()
+# print out the message according to the level
+# with timestamp
+#
+# usage:
+# log "$LOG_LEVEL" "$message" ($LOG_LEVEL = DEBUG|INFO|WARNING|ERROR)
```
Comment (sc979) [inline bin/registerServerTopology.sh:129]: DEBUG seems never usd

File: bin/registerServerTopology.sh [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+# log "DEBUG" "This is a DEBUG_LOG_LEVEL message"
+# log "INFO" "This is a INFO_LOG_LEVEL message"
+#
+function log() {
+  TIMESTAMP=$(date --rfc-3339=seconds)
+
+  if [[ -z "${1}" || -z "${2}" ]]
+  then
+    echo "${TIMESTAMP} - ERROR : Missing argument"
+    echo "${TIMESTAMP} - ERROR : Usage log \"INFO\" \"Message log\" "
```
Comment (sc979) [inline bin/registerServerTopology.sh:141]:
```suggestion
    echo "${TIMESTAMP} - ERROR : Usage log \"TYPE\" \"Message log\" "
```

File: bin/registerServerTopology.sh [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+  PARSED_URL[SCHEME]="$(echo $1 | grep :// | sed -e's,^\(.*://\).*,\1,g')"
+  # remove the protocol
+  url="$(echo ${1/${PARSED_URL[SCHEME]}/})"
+  # extract the user (if any)
+  userpass="$(echo $url | grep @ | cut -d@ -f1)"
+  pass="$(echo $userpass | grep : | cut -d: -f2)"
+  if [ -n "$pass" ]; then
+    user="$(echo $userpass | grep : | cut -d: -f1)"
+  else
+      user=$userpass
```
Comment (sc979) [inline bin/registerServerTopology.sh:173]: indentation

File: bin/registerServerTopology.sh [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+  hostname: ${HOSTNAME}
+  type: ${CURRENT_NODE_TYPE}
+  address: ${CURRENT_NODE_ADDRESS}
+
+EOD
+
+  echo 'Do you want to register this server with those information? (y/n) '
+  read -r IS_VALID
+
+  if [[ $IS_VALID != 'y' ]];
```
Comment (sc979) [inline bin/registerServerTopology.sh:285]:
```suggestion
  if [[ $IS_VALID !=~ ^y(es)?$ ]];
```

File: bin/registerServerTopology.sh [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+    exit 1
+  fi
+}
+#========= end of function set_variable()
+
+
+#========= begin of function get_api_token()
+# Get the X-AUTH-TOKEN used in register request
+function get_api_token() {
+  API_RESPONSE=$(curl -s -X POST ${INSECURE:+--insecure} -H "Content-Type: application/json" \
```
Comment (sc979) [inline bin/registerServerTopology.sh:109]:
```suggestion
  API_RESPONSE=$(curl -s -X POST "${INSECURE:+--insecure}" -H "Content-Type: application/json" \
```

File: bin/registerServerTopology.sh [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+
+  if [[ $HTTP_CODE == "201" ]];
+  then
+    log "INFO" "The CURRENT NODE ${CURRENT_NODE_TYPE}: '${CURRENT_NODE_NAME}@${CURRENT_NODE_ADDRESS}' linked to TARGET NODE: ${TARGET_NODE_ADDRESS} has been added"
+  elif [[ $RESPONSE_MESSAGE != "" ]];
+  then
+    log "ERROR" "${RESPONSE_MESSAGE}"
+    exit 1
+  else
+    log "ERROR" "An error occurred while contacting the API using: '${TARGET_NODE_ADDRESS}/centreon/api/latest/platform/topology' "
```
Comment (kduret) [inline bin/registerServerTopology.sh:312]:
```suggestion
    log "ERROR" "An error occurred while contacting the API using: '${TARGET_NODE_ADDRESS}/${ROOT_CENTREON_FOLDER}/api/latest/platform/topology' "
```

File: bin/registerServerTopology.sh [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+#========= begin of register_server()
+
+
+#========= begin of set_variable_from_template()
+function set_variable_from_template() {
+
+  source "$1"
+
+  if [[ ! $API_USERNAME \
+    || ! $CURRENT_NODE_TYPE \
```
Comment (kduret) [inline bin/registerServerTopology.sh:325]:
```suggestion
  if [[ ! $API_USERNAME \
    || ! API_TARGET_PASSWORD \
    || ! $CURRENT_NODE_TYPE \
```

File: bin/registerServerTopology.sh [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+    -d "${PAYLOAD}" \
+    "${TARGET_NODE_ADDRESS}/${ROOT_CENTREON_FOLDER}/api/latest/platform/topology" | grep -E "(HTTP/|message)"))
+
+  HTTP_CODE="$(echo ${API_RESPONSE[0]} | cut -d ' ' -f2)"
+  RESPONSE_MESSAGE=${API_RESPONSE[1]}
+
+  if [[ $HTTP_CODE == "201" ]];
+  then
+    log "INFO" "The CURRENT NODE ${CURRENT_NODE_TYPE}: '${CURRENT_NODE_NAME}@${CURRENT_NODE_ADDRESS}' linked to TARGET NODE: ${TARGET_NODE_ADDRESS} has been added"
+  elif [[ $RESPONSE_MESSAGE != "" ]];
```
Comment (sc979) [inline bin/registerServerTopology.sh:307]:
```suggestion
  elif [[ -z $RESPONSE_MESSAGE ]];
```

File: bin/registerServerTopology.sh [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+
+  source "$1"
+
+  if [[ ! $API_USERNAME \
+    || ! $CURRENT_NODE_TYPE \
+    || ! $TARGET_NODE_ADDRESS \
+    || ! $CURRENT_NODE_NAME \
+    || ! $CURRENT_NODE_ADDRESS \
+  ]]; then
+    log "ERROR" "Missing Parameters: please fill all the template's mandatory fields"
```
Comment (sc979) [inline bin/registerServerTopology.sh:330]:
```suggestion
    log "ERROR" "Missing Parameters: please fill all the template's mandatory parameters"
```

File: bin/registerServerTopology.sh [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+    exit 0
+  fi
+}
+#========= end of prepare_register_payload()
+
+
+#========= begin of register_server()
+# Send the request to register the server
+function register_server() {
+  IFS=$'\n' API_RESPONSE=($(curl -s -X POST ${INSECURE:+--insecure} -i -H "Content-Type: application/json" -H "X-AUTH-TOKEN: ${API_TOKEN}" \
```
Comment (sc979) [inline bin/registerServerTopology.sh:297]:
```suggestion
  IFS=$'\n' API_RESPONSE=($(curl -s -X POST "${INSECURE:+--insecure}" -i -H "Content-Type: application/json" -H "X-AUTH-TOKEN: ${API_TOKEN}" \
```

File: bin/registerServerTopology.sh [hunk 15]
```diff
@@ -0,0 +1,373 @@
...
+    log "ERROR" "An error occurred while contacting the API using: '${TARGET_NODE_ADDRESS}/centreon/api/latest/platform/topology' "
+    exit 1
+  fi
+}
+#========= begin of register_server()
+
+
+#========= begin of set_variable_from_template()
+function set_variable_from_template() {
+
```
Comment (sc979) [inline bin/registerServerTopology.sh:321]:
```suggestion
```

File: bin/registerServerTopology.sh (hunk 14 above)
Comment (sc979) [inline bin/registerServerTopology.sh:297]: $IFS is a special shell variable, you should rename it :
https://bash.cyberciti.biz/guide/$IFS

Comment (sc979) [reply bin/registerServerTopology.sh:297]: as discussed, overriding the IFS is required -> so you should explain the intention by adding a comment

Comment (jeremyjaouen) [reply bin/registerServerTopology.sh:297]: without IFS:

2020-12-16 09:54:57+01:00 - ERROR - 409

with:

2020-12-16 09:43:26+01:00 - ERROR - {"message":"A platform using the name : 'xxxx' or address : 'xxx.xx.x.x' already exists"}

I'll comment this portion of code to explain it

**Review by sc979 — APPROVED** [review]

File: bin/registerServerTopology.sh [hunk 16]
```diff
@@ -0,0 +1,517 @@
+#!/bin/bash
+
+# Set value empty to prevent already set value while re executing the script
```
Comment (sc979) [inline bin/registerServerTopology.sh:3]:
```suggestion
# Set value empty to prevent already set value while re-executing the script
```

**Review by sc979 — CHANGES_REQUESTED** [review]

**Review by sc979 — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Pull request merged**